### PR TITLE
Working GUID

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @andrewconner 
Not fully happy yet, there's still some boilerplate in there

### CODE REVIEW

**Comment by andrewconner** [comment]
Will go through this later today.

File: kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  def apply(headerCode: Short): GlobalIdCompanion = byHeaderCode(headerCode)
+}
+
+
+sealed trait ModelWithGlobalId {
+  type G <: ModelWithGlobalId
+  type M <: G
+  def id: Option[Id[M]]
+  def globalId(implicit header: TypeHeader[G]): Option[GlobalId] = id.map(GlobalId[G, M](_))
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: Why an implicit header instead of having an abstract field `header` in the trait?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: The headers are defined in companion objects, this way:
- They are available when you have the type around, but no instance (e.g. validating an id globalId.toId[User])
- They can be gathered and validated (for uniqueness) by reflection

Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: Oh, makes sense. I guess what I was really concerned about it the implicit pass through, which can make the code hard to follow. How about instead of this line something like
`def globalId(implicit header: TypeHeader[G]): Option[GlobalId] = id.map(GlobalId[G, M](_, header))`
and then line 23 (`apply` on the `GlobalId` object) turns into
`def apply[G <: ModelWithGlobalId, M <: G](id: Id[M], header: TypeHeader[G])`.
This compiles basically into the same thing (minus curried arguments) and is much easier to read, imho.

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: Sure I'll reduce the amount of implicit magic :)

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: What about:
`def apply[G <: ModelWithGlobalId, M <: G](id: Id[M])(implicit header: TypeHeader[G])`

Then it can be passed in explicitly or implicitly, and the API separates concerns.

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: +1

2014-03-27 11:28 GMT-07:00 Andrew Conner notifications@github.com:

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: But then there's no more Context bound, which is not good for my job
security.

2014-03-27 11:29 GMT-07:00 Léo Grimaldi leo.grimaldi.fr@gmail.com:

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: If you add a type lambda, you're good again.

File: kifi-backend/modules/common/app/com/keepit/model/User.scala [hunk 2]
```diff
@@ -24,7 +24,7 @@ case class User(
   userPictureId: Option[Id[UserPicture]] = None,
   seq: SequenceNumber[User] = SequenceNumber.ZERO,
   primaryEmailId: Option[Id[EmailAddress]] = None
-) extends ModelWithExternalId[User] with ModelWithState[User] with ModelWithSeqNumber[User]{
+) extends ModelWithExternalId[User] with ModelWithState[User] with ModelWithSeqNumber[User] with UserWithGlobalId {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/User.scala]: Only tangential to what you did... it's getting cake-y, and non-DRY with all the `[User]`s. Curious if there's a type trick where each of these can take a `self: M =>` and enforce that contract without an explicit annotation on the trait.

File: kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+sealed trait ModelWithGlobalId[M <: ModelWithGlobalId[M]] { self: Model[M] =>
+  type G <: ModelWithGlobalId[M]
+  def id: Option[Id[M]]
+  def globalId(implicit header: TypeHeader[M]): Option[GlobalId] = id.map(GlobalId[M](_))
+}
+
+trait UserWithGlobalId extends ModelWithGlobalId[User] { self: User => type G = UserWithGlobalId; }
+
+case object UserWithGlobalId extends GlobalIdCompanion {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: Hm, how will this work when we want a global ID of an object that's not in the common project?

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/db/ModelWithGlobalId.scala]: ie, social user data may be in a non-common project, but it's shared across multiple services. Would mean we couldn't use this.
